resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -1416,4 +1389,136 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
             }
         }
     }
+
+    private interface DnsResolveChannelProvider {
+        /**
+         * Return the next {@link ChannelFuture} that contains the {@link Channel} that should be used for resolving
+         * a chain of queries.
+         *
+         * @return  the {@link ChannelFuture}
+         */
+        ChannelFuture nextResolveChannel();
+
+        /**
+         * Called once resolve operations completes.
+         *
+         * @param resolveChannelFuture  the {@link ChannelFuture} that was previous obtained
+         *                              via {@link #nextResolveChannel()}.
+         * @param resolveFuture         the {@link Future} that holds the result of the resolve operation.
+         */
+        void resolveComplete(ChannelFuture resolveChannelFuture, Future<?> resolveFuture);
+
+        /**
+         * Close the {@link DnsResolveChannelProvider} and so cleanup resources if needed.
+         */
+        void close();
+    }
+
+    private static final class DnsResolveSameChannelProvider implements DnsResolveChannelProvider {
+
+        private final ChannelFuture resolveChannelFuture;
+
+        DnsResolveSameChannelProvider(Bootstrap bootstrap, SocketAddress localAddress) {
+            SocketAddress address = localAddress == null ? new InetSocketAddress(0) : localAddress;
+            resolveChannelFuture = bootstrap.bind(address);
+        }
+
+        @Override
+        public ChannelFuture nextResolveChannel() {
+            return resolveChannelFuture;
+        }
+
+        @Override
+        public void resolveComplete(ChannelFuture resolveChannelFuture, Future<?> resolveFuture) {
+            // Do nothing
+        }
+
+        @Override
+        public void close() {
+            resolveChannelFuture.channel().close();
+        }
+    }
+
+    private static final class DnsResolveSameChannelUntilTimeoutProvider implements DnsResolveChannelProvider {
+        private final Bootstrap bootstrap;
+        private final SocketAddress localAddress;
+        private volatile ChannelFuture resolveChannelFuture;
+
+        DnsResolveSameChannelUntilTimeoutProvider(Bootstrap bootstrap, SocketAddress localAddress) {
+            this.bootstrap = bootstrap;
+            this.localAddress = localAddress;
+        }
+
+        @Override
+        public ChannelFuture nextResolveChannel() {
+            if (resolveChannelFuture == null) {
```

Comment:
Is it possible to guarantee that this method is only invoked from event loop to avoid `volatile`? 
If not, we need to add another protection to guarantee we always create only one `resolveChannelFuture` at a time. Currently it feels racy. If 2 threads see this as `null` they will bind 2 different channels and then `close()` will close only the last one that wins.

Reply:
This was only done because close() might be called from a different thread … that said we could just schedule it on the loop